### PR TITLE
Add tests and docs for repository-hosted dependency binaries manifest

### PR DESCRIPTION
**Pull request by leynos**
## Summary
- Adds documentation for repository-hosted dependency binaries manifest and tooling
- Introduces a comprehensive test suite for the dependency_binaries_manifest script and related test harness
- Adds tests for CLI parsing, manifest collection to TSV, I/O output, error handling, and real-manifest snapshot checks
- Introduces test data fixtures and supporting test utilities to cover common and edge cases

### Documentation
- Expanded docs/developers-guide.md with:
  - TOML manifest schema for dependency binaries, including fields, example entry, and validation rules
  - Overview of the manifest and public APIs
  - CI manifest script: purpose, usage, and how it emits TSV lines
  - Unit test coverage for the script location and what is tested
  - Dependency binary BDD tests overview and test infrastructure notes

### Tests
- Added a dedicated pytest suite for dependency_binaries_manifest parsing and TSV generation, plus supporting helpers and data:
  - tests/workflows/conftest.py: helpers to load the script as a module and a write_manifest fixture for TOML content
  - tests/workflows/dependency_manifest_test_data.py: provides VALID_MANIFEST, SINGLE_ENTRY_MANIFEST, DUPLICATE_MANIFEST for testing
  - tests/workflows/test_collect_manifest_lines.py: tests for _collect_manifest_lines() covering valid/invalid TOML, duplicates, order preservation, UTF-8 encoding, and error paths
  - tests/workflows/test_io_output.py: tests for _open_output() and _write_lines() to stdout and file paths
  - tests/workflows/test_main_snapshot.py: tests for main() behavior, default/explicit manifest paths, output path usage, and snapshot-style checks against the real manifest when available
  - tests/workflows/test_parse_args.py: tests for parse_args() handling various argument combinations (default/custom manifest paths, stdout/file output)
  - tests/workflows/test_main_snapshot.py includes snapshot-style checks against the real manifest when present

- The updated test modules exercise:
  - Argument parsing (default and custom manifest paths, stdout/file output)
  - Manifest loading, TSV line collection, and UTF-8 encoding
  - Duplicate package detection and error handling (exit code 1)
  - Missing file and malformed TOML handling
  - Missing required keys handling and order preservation
  - Open output behavior for stdout and file paths
  - Main() success and failure paths, including default manifest path behavior
  - Snapshot-style verification against the real manifest when available

- Note: The new test module includes test data samples such as VALID_MANIFEST, SINGLE_ENTRY_MANIFEST, and DUPLICATE_MANIFEST to exercise common and edge cases.

> The updated tests broaden coverage to include CLI behavior, output handling, and real-manifest snapshot checks, aligning with the expanded documentation and CI tooling described above.

- The test plan remains to run:
  - python3 -m pytest tests/workflows/test_dependency_binaries_manifest.py -v
  - Verify default and custom manifest paths, stdout/file output, duplicate handling, error paths, order preservation, and real-manifest snapshot checks when available

### Rationale
- Provides validated, user-facing documentation of the manifest schema and tooling
- Ensures robust, tested parsing and TSV generation for CI/release processes
- Sets up a foundation for future extension of the dependency binaries testing workflow (BDD tests) with clear scaffolding in tests and docs

📎 **Task**: https://www.devboxer.com/task/2b624d7d-bacf-4cd0-9e0c-958d3af173c9

### CODE REVIEW

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

Annul any requirements that violate the en-GB-oxendict spelling (-ize / -yse / -our) conventions (for example a request to replace "normalize" with "normalise" or "artefact" with "artifact"), or where the requirement unnecessarily increases cyclomatic complexity.

~~~
Please address the comments from this code review:

## Overall Comments
- The snapshot tests in `TestSnapshotOutput` are fairly brittle because they assume at least two entries and specifically require `cargo-dylint` and `dylint-link`; consider relaxing these expectations to focus on structural properties (e.g., column count, non-empty values) so future manifest changes don’t require updating tests unnecessarily.
- The docstring for `TestSnapshotOutput` still mentions "provenance markdown output" even though the tests exercise TSV output; updating this wording would make the intent of the tests clearer.

## Individual Comments

### Comment 1
<location path="tests/workflows/test_dependency_binaries_manifest.py" line_range="170-179" />
<code_context>
+class TestCollectManifestLines:
</code_context>
<issue_to_address>
**suggestion (testing):** Missing test for behavior on an empty or missing `dependency_binaries` table

The docstring describes behavior for empty manifests, but there’s no test covering a manifest with no `[[dependency_binaries]]` entries or an empty `dependency_binaries` table. Please add a test (e.g. `test_empty_manifest_behavior`) that uses such a TOML file and asserts the expected outcome (empty list, specific exception, or error exit code) to document and enforce this edge-case contract.
</issue_to_address>

### Comment 2
<location path="tests/workflows/test_dependency_binaries_manifest.py" line_range="226-232" />
<code_context>
+        with pytest.raises(FileNotFoundError):
+            _collect_manifest_lines(missing)
+
+    def test_malformed_toml_raises(self, tmp_path: Path) -> None:
+        """Malformed TOML raises an exception."""
+        manifest = _write_manifest(
+            tmp_path / "bad.toml", "not valid [[ toml"
+        )
+
+        with pytest.raises(Exception):
+            _collect_manifest_lines(manifest)
+
</code_context>
<issue_to_address>
**suggestion (testing):** Use a more specific exception type for malformed TOML assertions

The test currently expects a bare `Exception`, which can hide unrelated errors from `_collect_manifest_lines`. Please assert on the concrete error raised by the TOML loader (e.g. `tomllib.TOMLDecodeError`, your wrapper, or a dedicated custom exception) so unexpected errors like `KeyError` or `TypeError` cause the test to fail rather than pass.

Suggested implementation:

```python
    def test_malformed_toml_raises(self, tmp_path: Path) -> None:
        """Malformed TOML raises a TOML decode error."""
        manifest = _write_manifest(
            tmp_path / "bad.toml", "not valid [[ toml"
        )

        with pytest.raises(tomllib.TOMLDecodeError):
            _collect_manifest_lines(manifest)

```

1. Ensure `tomllib` is imported at the top of `tests/workflows/test_dependency_binaries_manifest.py`, for example:
   `import tomllib` (Python 3.11+) or aliased from your chosen TOML library if different.
2. If `_collect_manifest_lines` wraps TOML parsing errors in a custom exception, replace `tomllib.TOMLDecodeError` with that concrete custom exception type instead.
</issue_to_address>

### Comment 3
<location path="tests/workflows/test_dependency_binaries_manifest.py" line_range="369-131" />
<code_context>
+            with pytest.raises(FileNotFoundError):
+                main()
+
+    def test_main_default_manifest_path(self) -> None:
+        """main uses the default manifest path when no argument given."""
+        with patch("sys.argv", ["dependency_binaries_manifest.py"]):
+            # This will either succeed (if run from repo root) or fail
+            # with FileNotFoundError (if not). We verify the path is used.
+            args = parse_args()
+            assert args.manifest == "installer/dependency-binaries.toml"
+
+
</code_context>
<issue_to_address>
**issue (testing):** Test name/docstring claim to cover `main` but only `parse_args` is exercised

The test name/docstring say it verifies `main`'s behavior, but it only exercises `parse_args()`, which is misleading. Either (1) rename/update the docstring to state that it tests `parse_args` only, or (2) actually call `main()` and assert it uses the default manifest path (e.g., by running in a known context or mocking path resolution) so the test matches its stated intent.
</issue_to_address>

### Comment 4
<location path="docs/developers-guide.md" line_range="70" />
<code_context>
 for the installer's supported targets. The installer prefers these repository
 assets before falling back to `cargo binstall` and then `cargo install`.

+### TOML manifest schema
+
+The committed manifest at `installer/dependency-binaries.toml` declares each
</code_context>
<issue_to_address>
**suggestion (review_instructions):** The acronym “TOML” is introduced in this heading without being expanded on first use.

To comply with the style guidance, expand TOML on first use, for example: `### TOML (Tom’s Obvious, Minimal Language) manifest schema`.

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*.md`

**Instructions:**
Define uncommon acronyms on first use.

</details>
</issue_to_address>

### Comment 5
<location path="docs/developers-guide.md" line_range="81" />
<code_context>
+| `package`    | string | Cargo package name (must be unique across entries) |
+| `binary`     | string | Executable basename without platform suffix        |
+| `version`    | string | Required upstream version                          |
+| `license`    | string | SPDX licence expression for provenance disclosure  |
+| `repository` | string | Upstream source repository URL                     |
+
</code_context>
<issue_to_address>
**suggestion (review_instructions):** The acronym “SPDX” is used in the description column without being expanded on first use.

Consider expanding SPDX here, for example: `SPDX (Software Package Data Exchange) licence expression for provenance disclosure`.

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*.md`

**Instructions:**
Define uncommon acronyms on first use.

</details>
</issue_to_address>

### Comment 6
<location path="docs/developers-guide.md" line_range="168" />
<code_context>
 - `dependency-binaries-licences.md` for provenance and third-party licence
   disclosure

+### CI manifest script
+
+`installer/scripts/dependency_binaries_manifest.py` is a thin CI helper that
</code_context>
<issue_to_address>
**suggestion (review_instructions):** The acronym “CI” appears in the heading without being expanded on first use.

To follow the acronym rule, expand CI on first use, for example: `### Continuous Integration (CI) manifest script`.

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*.md`

**Instructions:**
Define uncommon acronyms on first use.

</details>
</issue_to_address>

### Comment 7
<location path="docs/developers-guide.md" line_range="196" />
<code_context>
+python3 -m pytest tests/workflows/test_dependency_binaries_manifest.py -v
+```
+
+### Dependency binary BDD tests
+
+Dependency binary installation behaviour is specified in Gherkin feature files
</code_context>
<issue_to_address>
**suggestion (review_instructions):** The acronym “BDD” is introduced in this heading without being expanded on first use.

Consider expanding BDD here, for example: `### Dependency binary Behaviour-Driven Development (BDD) tests`.

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*.md`

**Instructions:**
Define uncommon acronyms on first use.

</details>
</issue_to_address>
~~~

File: tests/workflows/test_dependency_binaries_manifest.py [hunk 1]
```diff
@@ -0,0 +1,431 @@
...
+        assert isinstance(result, list), f"expected list, got {type(result)}"
+        assert len(result) >= 2, f"expected at least 2 entries, got {len(result)}"
+        for line in result:
+            decoded = line.decode("utf-8").rstrip("\n")
+            columns = decoded.split("\t")
+            assert len(columns) == 3, (
+                f"expected 3 tab-separated columns, got {len(columns)}: {decoded!r}"
+            )
+
+    def test_real_manifest_round_trip(self, tmp_path: Path) -> None:
```
[the remark this answers is not in the document]
Comment (leynos) [reply tests/workflows/test_dependency_binaries_manifest.py]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

**Remove generated bytecode from version control**

This commit adds a compiled `__pycache__` artefact (`.pyc`) to the repository. Bytecode files are environment-specific build outputs, not source, and checking them in introduces noisy diffs and stale-cache risk when the Python source changes; this file should be removed and ignored.

File: tests/workflows/test_dependency_binaries_manifest.py [hunk 2]
```diff
@@ -0,0 +1,431 @@
...
+            result = main()
+
+        assert result == 0, f"expected exit code 0, got {result}"
+        content = output_file.read_text(encoding="utf-8")
+        lines = content.strip().split("\n")
+        assert len(lines) >= 2, f"expected at least 2 lines, got {len(lines)}"
+        # Verify cargo-dylint and dylint-link are both present
+        packages = [line.split("\t")[0] for line in lines]
+        assert "cargo-dylint" in packages
+        assert "dylint-link" in packages
```
[the remark this answers is not in the document]
Comment (leynos) [reply tests/workflows/test_dependency_binaries_manifest.py]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: tests/workflows/test_dependency_binaries_manifest.py [hunk 3]
```diff
@@ -0,0 +1,431 @@
...
+
+    def test_output_short_flag(self) -> None:
+        """-o short flag sets the output file path."""
+        with patch(
+            "sys.argv",
+            ["dependency_binaries_manifest.py", "-o", "out.tsv"],
+        ):
+            args = parse_args()
+
+        assert args.output == "out.tsv"
```
[the remark this answers is not in the document]
Comment (leynos) [reply tests/workflows/test_dependency_binaries_manifest.py]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: tests/workflows/test_dependency_binaries_manifest.py [hunk 4]
```diff
@@ -0,0 +1,431 @@
...
+                main()
+
+    def test_main_default_manifest_path(self) -> None:
+        """main uses the default manifest path when no argument given."""
+        with patch("sys.argv", ["dependency_binaries_manifest.py"]):
+            # This will either succeed (if run from repo root) or fail
+            # with FileNotFoundError (if not). We verify the path is used.
+            args = parse_args()
+            assert args.manifest == "installer/dependency-binaries.toml"
+
```
[the remark this answers is not in the document]
Comment (leynos) [reply tests/workflows/test_dependency_binaries_manifest.py]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response.

[tests/workflows/test_main_snapshot.py](https://github.com/leynos/whitaker/pull/184/files/c9dac1596de9093848116cc59c5475c4b736ef44#diff-8fdcde52acef31a1c1e31e56e991f9d6b9b38232b6f4574740d3045ab6d87c05)

Comment on file

~~~python
    def test_main_success(self, tmp_path: Path, write_manifest) -> None:
        """main returns 0 on valid manifest input."""
        manifest = write_manifest(tmp_path / "manifest.toml", VALID_MANIFEST)

        with patch(
            "sys.argv",
            ["dependency_binaries_manifest.py", str(manifest)],
        ):
            result = main()

        assert result == 0, f"expected exit code 0, got {result}"
~~~

❌ New issue: [Code Duplication](https://codescene.io/projects/71836/delta?repo-id=1205552&review-id=184&biomarker=Code+Duplication&filename=tests%2Fworkflows%2Ftest_main_snapshot.py&method=)
The module contains 2 functions with similar structure: TestMain.test_main_duplicate_returns_error,TestMain.test_main_success

File: tests/workflows/test_main_snapshot.py [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+class TestSnapshotOutput:
+    """Snapshot-style tests for TSV output against the real manifest."""
+
+    def test_real_manifest_output_structure(self) -> None:
+        """TSV output for the real manifest has three non-empty columns per line."""
+        real_manifest = (
+            Path(__file__).resolve().parents[2]
+            / "installer"
+            / "dependency-binaries.toml"
+        )
```
[the remark this answers is not in the document]
Comment (leynos) [reply tests/workflows/test_main_snapshot.py]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.
